pkg/store/proxy_heap.go

```diff
@@ -733,11 +799,73 @@ func (l *eagerRespSet) Labelset() string {
 	return labelpb.PromLabelSetsToString(l.st.LabelSets())
 }
 
+func (l *eagerRespSet) HasMultiReplicaSeries() bool {
+	for _, m := range l.seenReplicas {
+		if len(m) > 1 {
+			return true
+		}
+	}
+	return false
+}
+
 type respSet interface {
 	Close()
-	At() *storepb.SeriesResponse
+	At() *signedResponse
 	Next() bool
 	StoreID() string
 	Labelset() string
 	Empty() bool
+	HasMultiReplicaSeries() bool
+}
+
+// sortLabelsForDedup reorders the series so that replica labels are at the end.
+// Labels used for query pushdown are placed right before deduplication labels.
+// sortLabelsForDedup will also add replica labels and values to the seenReplicas map.
+// It returns a signedResponse containing the series response and the labels hash without
+// replica labels.
+func sortLabelsForDedup(
+	resp *storepb.SeriesResponse,
+	replicaLabels []string,
+	replicaLabelSet map[string]struct{},
+	seenReplicas map[string]map[string]struct{},
+	buf []byte,
+) *signedResponse {
+	s := resp.GetSeries()
+	if s == nil {
+		return newUnsignedResponse(resp)
+	}
+	if len(replicaLabels) == 0 {
+		signature := labelpb.HashWithPrefix("", s.Labels)
+		return newSignedResponse(signature, resp)
+	}
+
+	for _, lbl := range s.Labels {
+		if _, ok := replicaLabelSet[lbl.Name]; !ok {
+			continue
+		}
+		if _, ok := seenReplicas[lbl.Name]; !ok {
+			seenReplicas[lbl.Name] = make(map[string]struct{})
+		}
+		seenReplicas[lbl.Name][lbl.Value] = struct{}{}
+	}
+
+	sort.Slice(s.Labels, func(i, j int) bool {
```

Comment:
This is actually still expensive, but as next steps we can push it down to the stores since replica labels were added to the `SeriesRequest`

<img width="1881" alt="image" src="https://user-images.githubusercontent.com/1286231/191709797-9ee2e136-10c9-4c2c-91f0-2d5086a1c31b.png">